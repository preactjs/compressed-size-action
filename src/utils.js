import fs from 'fs';
import prettyBytes from 'pretty-bytes';

/**
 * Check if a given file exists and can be accessed.
 * @param {string} filename
 */
export async function fileExists(filename) {
	try {
		await fs.promises.access(filename, fs.constants.F_OK);
		return true;
	} catch (e) {}
	return false;
}

/**
 * Remove any matched hash patterns from a filename string.
 * @param {string=} regex
 * @returns {(((fileName: string) => string) | undefined)}
 */
export function stripHash(regex) {
	if (regex) {
		console.log(`Stripping hash from build chunks using '${regex}' pattern.`);
		return function (fileName) {
			return fileName.replace(new RegExp(regex), (str, ...hashes) => {
				hashes = hashes.slice(0, -2).filter((c) => c != null);
				if (hashes.length) {
					for (let i = 0; i < hashes.length; i++) {
						const hash = hashes[i] || '';
						str = str.replace(hash, hash.replace(/./g, '*'));
					}
					return str;
				}
				return '';
			});
		};
	}

	return undefined;
}

/**
 * @param {number} delta
 * @param {number} originalSize
 */
export function getDeltaText(delta, originalSize) {
	let deltaText = (delta > 0 ? '+' : '') + prettyBytes(delta);
	if (Math.abs(delta) === 0) {
		// only print size
	} else if (originalSize === 0) {
		deltaText += ` (new file)`;
	} else if (originalSize === -delta) {
		deltaText += ` (removed)`;
	} else {
		const percentage = Number(((delta / originalSize) * 100).toFixed(2));
		deltaText += ` (${percentage > 0 ? '+' : ''}${percentage}%)`;
	}
	return deltaText;
}

/**
 * @param {number} delta
 * @param {number} originalSize
 */
export function iconForDifference(delta, originalSize) {
	if (originalSize === 0) return 'üÜï';

	const percentage = Math.round((delta / originalSize) * 100);
	if (percentage >= 50) return 'üÜò';
	else if (percentage >= 20) return 'üö®';
	else if (percentage >= 10) return '‚ö†Ô∏è';
	else if (percentage >= 5) return 'üîç';
	else if (percentage <= -50) return 'üèÜ';
	else if (percentage <= -20) return 'üéâ';
	else if (percentage <= -10) return 'üëè';
	else if (percentage <= -5) return '‚úÖ';
	return '';
}

/**
 * Create a Markdown table from text rows
 * @param {string[][]} rows
 */
function markdownTable(rows) {
	if (rows.length == 0) {
		return '';
	}

	// Skip all empty columns
	while (rows.every(columns => !columns[columns.length - 1])) {
		for (const columns of rows) {
			columns.pop();
		}
	}

	const [firstRow] = rows;
	let columnLength = firstRow.length;

	// Hide `Change` column if they are all `0 B`
	if (columnLength === 3 && rows.every(columns => columns[2] === '0 B')) {
		columnLength -= 1;
		for (const columns of rows) {
			columns.pop();
		}
	}

	if (columnLength === 0) {
		return '';
	}

	return [
		// Header
		['Filename', 'Size', 'Change', ''].slice(0, columnLength),
		// Align
		[':---', ':---:', ':---:', ':---:'].slice(0, columnLength),
		// Body
		...rows
	].map(columns => `| ${columns.join(' | ')} |`).join('\n');
}

/**
 * @typedef {Object} Diff
 * @property {string} filename
 * @property {number} size
 * @property {number} delta
 */

/**
 * Create a Markdown table showing diff data
 * @param {Diff[]} files
 * @param {object} options
 * @param {boolean} [options.showTotal]
 * @param {boolean} [options.collapseUnchanged]
 * @param {boolean} [options.omitUnchanged]
 * @param {number} [options.minimumChangeThreshold]
 */
export function diffTable(files, { showTotal, collapseUnchanged, omitUnchanged, minimumChangeThreshold }) {
	let changedRows = [];
	let unChangedRows = [];

	let totalSize = 0;
	let totalDelta = 0;
	for (const file of files) {
		const { filename, size, delta } = file;
		totalSize += size;
		totalDelta += delta;

		const originalSize = size - delta;
		const isUnchanged = Math.abs(delta) < minimumChangeThreshold;

		if (isUnchanged && omitUnchanged) continue;

		const columns = [
			`\`${filename}\``,
			prettyBytes(size),
			getDeltaText(delta, originalSize),
			iconForDifference(delta, originalSize)
		];
		if (isUnchanged && collapseUnchanged) {
			unChangedRows.push(columns);
		} else {
			changedRows.push(columns);
		}
	}

	let out = markdownTable(changedRows);

	if (unChangedRows.length !== 0) {
		const outUnchanged = markdownTable(unChangedRows);
		out += `\n\n<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>\n\n${outUnchanged}\n\n</details>\n\n`;
	}

	if (showTotal) {
		const totalOriginalSize = totalSize - totalDelta;
		let totalDeltaText = getDeltaText(totalDelta, totalOriginalSize);
		let totalIcon = iconForDifference(totalDelta, totalOriginalSize);
		out = `**Total Size:** ${prettyBytes(totalSize)}\n\n${out}`;
		out = `**Size Change:** ${totalDeltaText} ${totalIcon}\n\n${out}`;
	}

	return out;
}

/**
 * Convert a string "true"/"yes"/"1" argument value to a boolean
 * @param {string} v
 */
export function toBool(v) {
	return /^(1|true|yes)$/.test(v);
}
