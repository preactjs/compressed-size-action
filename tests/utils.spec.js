import { toBool, getDeltaText, iconForDifference, diffTable, fileExists, stripHash } from '../src/utils.js';

test('toBool', () => {
	expect(toBool('1')).toBe(true);
	expect(toBool('true')).toBe(true);
	expect(toBool('yes')).toBe(true);

	expect(toBool('0')).toBe(false);
	expect(toBool('false')).toBe(false);
	expect(toBool('no')).toBe(false);
});

test('getDeltaText', () => {
	expect(getDeltaText(5000, 20000)).toBe('+5 kB (+25%)');
	expect(getDeltaText(-5000, 20000)).toBe('-5 kB (-25%)');
	expect(getDeltaText(210, 0)).toBe('+210 B (new file)');
	expect(getDeltaText(0, 0)).toBe('0 B');
	expect(getDeltaText(4875, 20000)).toBe('+4.88 kB (+24.38%)');
	expect(getDeltaText(-4875, 20000)).toBe('-4.88 kB (-24.38%)');
});

test('iconForDifference', () => {
	expect(iconForDifference(0, 5000)).toBe('');
	expect(iconForDifference(5500, 5000)).toBe('🆘');
	expect(iconForDifference(-550, 5000)).toBe('👏');
});

test('diffTable', () => {
	const files = [
		{
			filename: 'one.js',
			size: 5000,
			delta: 2500
		},
		{
			filename: 'two.js',
			size: 5000,
			delta: -2500
		},
		{
			filename: 'three.js',
			size: 300,
			delta: 0
		},
		{
			filename: 'four.js',
			size: 4500,
			delta: 9
		}
	];
	const defaultOptions = {
		showTotal: true,
		collapseUnchanged: true,
		omitUnchanged: false,
		minimumChangeThreshold: 1
	};

	expect(diffTable(files, { ...defaultOptions })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, showTotal: false })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, collapseUnchanged: false })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, omitUnchanged: true })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, minimumChangeThreshold: 10 })).toMatchSnapshot();
	expect(diffTable(files.map(file => ({...file, delta: 0})), { ...defaultOptions })).toMatchSnapshot();

	expect(diffTable([files[2]], { ...defaultOptions })).toMatchSnapshot();
});

test('fileExists', async () => {
	expect(await fileExists('package.json')).toBe(true);
	expect(await fileExists('file-that-does-not-exist')).toBe(false);
});

test('stripHash', () => {
	expect(stripHash('\\b\\w{5}\\.')('foo.abcde.js')).toBe('foo.js');
	expect(stripHash('\\.(\\w{5})\\.chunk\\.js$')('foo.abcde.chunk.js')).toBe('foo.*****.chunk.js');
	expect(stripHash('')).toBe(undefined);
});
