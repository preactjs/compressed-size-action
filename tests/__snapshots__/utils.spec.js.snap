// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffTable 1`] = `
"**Size Change:** +9 B (+0.06%) 

**Total Size:** 14.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (+100%) | üÜò |
| \`two.js\` | 5 kB | -2.5 kB (-33.33%) | üéâ |
| \`four.js\` | 4.5 kB | +9 B (+0.2%) |  |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| \`three.js\` | 300 B |

</details>

"
`;

exports[`diffTable 2`] = `
"| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (+100%) | üÜò |
| \`two.js\` | 5 kB | -2.5 kB (-33.33%) | üéâ |
| \`four.js\` | 4.5 kB | +9 B (+0.2%) |  |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| \`three.js\` | 300 B |

</details>

"
`;

exports[`diffTable 3`] = `
"**Size Change:** +9 B (+0.06%) 

**Total Size:** 14.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (+100%) | üÜò |
| \`two.js\` | 5 kB | -2.5 kB (-33.33%) | üéâ |
| \`three.js\` | 300 B | 0 B |  |
| \`four.js\` | 4.5 kB | +9 B (+0.2%) |  |"
`;

exports[`diffTable 4`] = `
"**Size Change:** +9 B (+0.06%) 

**Total Size:** 14.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (+100%) | üÜò |
| \`two.js\` | 5 kB | -2.5 kB (-33.33%) | üéâ |
| \`four.js\` | 4.5 kB | +9 B (+0.2%) |  |"
`;

exports[`diffTable 5`] = `
"**Size Change:** +9 B (+0.06%) 

**Total Size:** 14.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (+100%) | üÜò |
| \`two.js\` | 5 kB | -2.5 kB (-33.33%) | üéâ |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`three.js\` | 300 B | 0 B |
| \`four.js\` | 4.5 kB | +9 B (+0.2%) |

</details>

"
`;

exports[`diffTable 6`] = `
"**Size Change:** 0 B 

**Total Size:** 14.8 kB



<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| \`one.js\` | 5 kB |
| \`two.js\` | 5 kB |
| \`three.js\` | 300 B |
| \`four.js\` | 4.5 kB |

</details>

"
`;

exports[`diffTable 7`] = `
"**Size Change:** 0 B 

**Total Size:** 300 B



<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| \`three.js\` | 300 B |

</details>

"
`;
